core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -217,26 +248,20 @@ public synchronized void noMoreTasks()
         checkInputFinished();
     }
 
-    private synchronized void checkInputFinished()
+    private void checkInputFinished()
     {
         if (failure != null) {
             return;
         }
 
-        if (inputFinished) {
+        if (outputSource != null) {
```

Comment:
check `outputReady` future instead? This would be more consistent with other checks in class.

Reply:
Future is unblocked asynchronously =\